### PR TITLE
Artist draw pattern backward

### PR DESCRIPTION
**Pull request by nkiruka**
Bug:  Artist does not draw any pattern backwards.  
pre-fix-pattern

Solution:  The following changes were implemented:

- Added  two variables for x and y coordinates for clipping the image to allow for future updates.

- Changed the x coordinate for clipping the image from  0 to image's width/2.  Prior to fix, x and y coordinates were (0, 0).  Therefore, if a distance of -50 px is selected, a blank space is clipped resulting in no image being rendered.  To resolve this, the start point for clipping the image is now set at the mid-point (image's width/2), which ensures the selection of backward or forward movement starts at the same point.

post-fix-pattern

### CODE REVIEW

**Comment by caleybrock** [comment]
Question for CSF - does validation rely on the exact output of the simulation? My understanding is that this might change the starting point of the pattern when moving forward?

**Comment by Hamms** [comment]
Validation does indeed depend on the exact output of the simulation (we actually do pixel-by-pixel matching), but we specifically compare this output to the result of running the solution code. So as long as the starting point of the pattern changes consistently for both user code and solution code, this should be fine.

**Comment by Hamms** [comment]
Test failure looks like a side effect of this branch being created from staging a while ago. If you update staging locally, then `git rebase staging` this branch it should be fine

**Comment by balderdash** [comment]
The rainbow pattern asset is 800px long, so I don't think this will work if you draw a diagonal across the whole canvas. I.e. if you jump to (0, 0), turn 45 degrees, and move forward 500 pixels, you'll run out of space in the pattern asset.

File: apps/src/turtle/turtle.js [hunk 1]
```diff
@@ -547,12 +547,15 @@ class Visualization {
...
+      const imgSourceY = 0;
+
       if (img.width !== 0) {
         this.ctxScratch.drawImage(img,
           // Start point for clipping image
-          0, 0,
+          imgSourceX, imgSourceY,
           // clip region size
-          distance + img.height / 2, img.height,
+          distance, img.height,
```
Comment (joshlory) [inline apps/src/turtle/turtle.js]: This PR is great!  I wonder if we should keep the nonsensical ` + img.height / 2` as-is for this PR, and open a follow up PR with that change?

**Comment by nkiruka** [comment]
Updated PR based on suggestions.  Tests also updated.  Please review.

File: apps/src/turtle/turtle.js [hunk 2]
```diff
@@ -228,6 +228,10 @@ class Visualization {
     this.turtleFrame_ = 0;
   }
 
+  clipImage() {
```
Comment (balderdash) [inline apps/src/turtle/turtle.js:231]: It looks like you didn't end up using this function, so you can remove it.

File: apps/src/turtle/turtle.js [hunk 3]
```diff
@@ -1510,7 +1524,7 @@ Artist.prototype.step = function (command, values, options) {
       result = this.calculateSmoothAnimate(options, distance);
       tupleDone = result.tupleDone;
       this.visualization.setHeading(heading);
-      this.visualization.jumpForward(result.distance);
+      this.jumpForward(result.distance);
```
Comment (balderdash) [inline apps/src/turtle/turtle.js:1527]: This changed with Josh's recent refactoring, it should be `this.visualization.jumpForward(...`

**Comment by nkiruka** [comment]
Oops, I didn't mean to merge PR.  I will revert back to previous PR.

**Pull request merged**